File: LeetCode_Algorithm/150(Evaluate Reverse Polish Notation)/150.js
var evalRPN = function(tokens) {
  // 문자열 to 숫자에서 == NaN 비교문을 사용하는 것이 아닌
  // isNaN함수를 사용하여 문제 해결해야 함
  // 67ms 73% 45.6MB 41%
  let stack = []
  for (let token of tokens) {
    if (!isNaN(token)) {
      stack.push(parseInt(token))
    } else {
      let num2 = stack.pop()
      let num1 = stack.pop()
      stack.push(cal(num1, token, num2))
    }
  }  
  return stack[stack.length - 1]
};
var cal = function(num1, operator, num2) {
  if (operator == '+')
    return (num1 + num2)
  else if (operator == '-')
    return (num1 - num2)
  else if (operator == '*')
    return (num1 * num2)
  else
    return parseInt(num1 / num2)
}

// console.log(parseInt("12") !== NaN)
console.log(isNaN(parseInt("+")))
console.log(isNaN(parseInt("12")))


console.log(evalRPN(["2","1","+","3","*"]))

File: LeetCode_Algorithm/150(Evaluate Reverse Polish Notation)/150.py
'''
LeetCode 150 : 

# sol1 67ms 64% 14.3MB 44%
작은 양의 정수 // 큰 음수 = -1
해결한 방법 int(작은 양의 정수 / 큰 음수)
'''
class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        stack = []
        for token in tokens:
            if token.isdigit() or token[1:].isdigit():
                stack.append(int(token))
            else:
                num2 = stack.pop()
                num1 = stack.pop()
                # print(num1, token, num2, self.cal(num1, token, num2))
                stack.append(self.cal(num1, token, num2))
        return stack[-1]
        
    def cal(self, num1, operator, num2):
        if operator == '+':
            return num1 + num2
        elif operator == '-':
            return num1 - num2
        elif operator == '*':
            return num1 * num2
        elif operator == '/':
            return int(num1 / num2)

# print(Solution().evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))
print(Solution().evalRPN(["4","-2","/","2","-3","-","-"]))

